### ISSUE TITLE
File system unresponsive when working with large quantities of files

### ISSUE THREAD

**Issue by bpadovese** [coastal-science/HALLO-annotation#3]
There is an issue in the file system when working with a large quantity of files. In my case I have 2000+ files, and when i select all (or a large quantity of files) and click "include" or "exclude", at first nothing seems to happen. It will take a while for the files to be included and sometimes i get this error: `A server/network error occurred. The server is not responding to the request, and may be restarted automatically`

If this error occurs, the remaining files are not included or excluded.

The biggest issue here however, is that there is no visual indication that some processing is occurring. So it seems like the system isnt doing anything. The first time this happened, I thought the system didnt get the command so i tried again and again and ended up crashing the app.

My suggestion would be to add a loading circle to the panel only, and disable the scan/verify/delete/include/exclude buttons while any file processing is happening. This way, the user knows the system is processing their request and they can move on to another tab to do something else.

**Comment by bpadovese** [comment]
Another enhancement related to the file system but which i think can extend to any checkbox component is the ability to shift select more than one box.

Currently, we can either select checkboxes one by one or select all of them by clicking select all.

To give a bit more detail imagine that we have 10 files. 

- User selects file 1
- User shift + click on file 5
- Files 1, 2, 3, 4 and 5 are selected. 

Same idea for deselect.

**Comment by bpadovese** [comment]
When going on Segments -> new -> and clicking on the dropdown list, I get a massive lag before the list is shown. This could possibly be because it is trying to display 2000+ files, but I am not sure.

**Comment by yue-su** [comment]
> 
The ability to mark files as included is actually just a way to add a tag to the file, to give the model developer some convenience when they need to handpick a few files that need to be processed. This feature was originally designed to handle small batches of files, so they are presented in a drop-down menu in other modules. It does not affect the functionality of other modules when requesting the file, for example, it is not necessary to pre-set the file as included when importing segments/annotations from csv.

Having thousands of files marked as included is a very different usage scenario, and in this case, it does not make sense to use the drop-down menu as a way to quickly select them anymore. In this user case, marking files as included is not necessary and this feature can be removed. The user interface could be tweaked accordingly, for example, by selecting the files in the file management interface and then generating segments directly.

As end-users,  you guys might need to discuss the use cases to decide whether we need to continue using the include/ exclude feature, as it really does not help improve productivity when including large numbers of files. I would like to have @fsfrazao weigh in on this issue and give some opinions before I move on to making changes.

**Comment by yue-su** [comment]
> Another enhancement related to the file system but which i think can extend to any checkbox component is the ability to shift select more than one checkbox.
> 
> Currently, we can either select checkboxes one by one or select all of them by clicking select all.
> 
> To give a bit more detail imagine that we have 10 files.
> 
> * User selects file 1
> * User shift + click on file 5
> * Files 1, 2, 3, 4 and 5 are selected.
> 
> Same idea for deselect.

This feature is available, I have tested it on Edge and Chrome on windows and it should be working. Maybe there are compatibility issues on some browsers, I don't know what version of browser you are using now?

**Comment by fsfrazao** [comment]
> > 
> 
> The ability to mark files as included is actually just a way to add a tag to the file, to give the model developer some convenience when they need to handpick a few files that need to be processed. This feature was originally designed to handle small batches of files, so they are presented in a drop-down menu in other modules. It does not affect the functionality of other modules when requesting the file, for example, it is not necessary to pre-set the file as included when importing segments/annotations from csv.
> 
> Having thousands of files marked as included is a very different usage scenario, and in this case, it does not make sense to use the drop-down menu as a way to quickly select them anymore. In this user case, marking files as included is not necessary and this feature can be removed. The user interface could be tweaked accordingly, for example, by selecting the files in the file management interface and then generating segments directly.
> 
> As end-users, you guys might need to discuss the use cases to decide whether we need to continue using the include/ exclude feature, as it really does not help improve productivity when including large numbers of files. I would like to have @fsfrazao weigh in on this issue and give some opinions before I move on to making changes.


I think your suggestion of creating the segments directly from the file management panel would be preferable.

**Comment by fsfrazao** [comment]
I can confirm that shift select feature work on my browser (Chrome 101.0.4951.54 (Official Build) (64-bit))

**Comment by bpadovese** [comment]
> > 
> 
> The ability to mark files as included is actually just a way to add a tag to the file, to give the model developer some convenience when they need to handpick a few files that need to be processed. This feature was originally designed to handle small batches of files, so they are presented in a drop-down menu in other modules. It does not affect the functionality of other modules when requesting the file, for example, it is not necessary to pre-set the file as included when importing segments/annotations from csv.
> 
> Having thousands of files marked as included is a very different usage scenario, and in this case, it does not make sense to use the drop-down menu as a way to quickly select them anymore. In this user case, marking files as included is not necessary and this feature can be removed. The user interface could be tweaked accordingly, for example, by selecting the files in the file management interface and then generating segments directly.
> 
> As end-users, you guys might need to discuss the use cases to decide whether we need to continue using the include/ exclude feature, as it really does not help improve productivity when including large numbers of files. I would like to have @fsfrazao weigh in on this issue and give some opinions before I move on to making changes.

I will let you guys decide what youu think is best regarding this. My main point here is that even though we are designing all these nice options for users to have an efficient working environment, I think the major use case would be pretty simple.

- User collected a bunch of files to annotate or detections to validate.
- create several segments to annotate/validate.
- I think the chances of having several hundreds or some thousands will be pretty common

So a way to accommodate this would be good.

**Comment by bpadovese** [comment]
> I can confirm that the shift select feature works on my browser (Chrome 101.0.4951.54 (Official Build) (64-bit))

You guys are right. It does work on chrome. It isnt working on firefox for some reason.

**Comment by yue-su** [comment]
@fsfrazao @bpadovese 

I agree with Brune that perhaps some of the user interface design could be simplified accordingly. I would do some optimization tweaks to make the workflow more concise. Here are some ideas:

- In the file management interface, remove the features of include and exclude, and remove the dropdown menu design.
- In the file management interface, users can select one or more files and then perform automatic generation of segments.
- Is it still necessary to keep the feature of adding a single segment manually? This feature doesn't feel like it will be used much, so I suggest removing it too.

**Comment by yue-su** [comment]
Guys, please help me understand the purpose of this app. In my understanding, it's more like a 'tool' rather than a 'system', so I was expecting this would be a use-it-and-leave-it tool, not for managing data-intensive computation or manipulation. The way we test it to handle more than 50k items it's actually more like testing the browser's capability, and network throughput, not the software itself. That's why it's freezing and timing out all the time. The way we test it it's more like testing a desktop application rather than a web-based tool. 

If a batch was supposed to be assigned to a real user, is it realistic to have thousands of segments in a batch for people to review? The UI interface is designed to handle dozens of segments per batch, as I think that is a reasonable number for real people to go through. Also, the segments, are reusable and can be added to multiple batches, and could be disposed of once the review is done.  I have tested it on the SFU instance, it can handle 10k files, 4-5k segments, and 4-5k annotations fairly ok, I didn't expect there would be 10 times more data per user. 

If that is indeed the use case, I would have to adjust the structure of the app from frontend heavy to backend heavy. Basically, it means all the data manipulation will happen on the backend in order to keep the frontend lightweight. But there are tradeoffs:

1. The frontend would be just a UI interface, as we can't cache too much data on the frontend, each time the user does something, it has to send or request the data from the backend, even if it's just a quick search, or parse a CSV file, which there would be significant delays. 
2. The system would have to scale up to handle this amount of data. I can no longer keep the simplicity. There will be pagination, dynamic data loading, and server-side rendering everywhere. It's going to be a mid-size team project. 

I can keep working on improving the user experience such as tweaking the UI or adding small features. However, it is hard to improve performance under the current structure design. I don't think it is worth it either, as I build this vehicle for running at 100km/h, not for 1000km/h.

**Comment by fsfrazao** [comment]
@yue-su, your assumptions for use cases are pretty much aligned with the context for which this app was developed (the HALLO project). Batches should be in the order of 10s or 100s of segments. I think most of the constraints Bruno is finding for his use case are justified for HALLO. It is important, however, that these constraints are known and made explicit, so that people can judge whether the application is suitable for their needs.

For example, Audacity and Raven are two (desktop) applications frequently used by bioacousticians and they have quite different sets of features. Audacity doesn't handle batches, for example. SO if someone has 100s of small files to inspect, they might find that Audacity is will not offer them a smooth workflow. If they Have fewer long files (a few 1h files as opposed to many 30s files), they might find that Audacity is perfectly fine. In most cases, you could probably do what you want in either application, you just might have to adjust your workflow to fit what the tool expects. But if you know how you have to work with each of these tools, you can decide how much you are willing to bend your ideal workflow.

So if there are somewhat clear constraints for the annotation app, it is helpful to make them known in several places (e.g.: a section in the github readme, in the help pages within the app, through error and warning messages like "Error: a batch can only contain 1000 segments" or " The number of segments in the system is approaching the maximum recommended limit. Keeping the number of segments below X will benefit response time and prevent crashes")

Bruno could probably work his workflow to use smaller batches.

In summary, constraints are fine and I think they are aligned with the APP's purpose, they just need to be made explicit.

**Comment by fsfrazao** [comment]
Regarding Bruno's points:

1-4: I think these are small feature adjustments. @yue-su what are your thoughts on each of these? Are these things you think are easy to implement and worth having in the HALLO version already or just leave it to the MAIPL version?

5- IThink it's fair to limit the number of segments that can be created at a time. Either with a hard limit (the app gives an error if the number of segments exceeds a limit) or with just a warning saying that it's not recommended and it might take a while, but the developer can choose to go on. Depends on whether the large number only cause a small inconvenience (like it takes a while but everything should work) or bigger problems (like, it will probably crash and not complete the creation of the segment) 

6- I think this might be more of a problem even in the context of HALLO. Even if you are not going to create large batches or add lot's of segments at a time, as time goes by the number of segments will continue to increase, right? So what's is the supported use here? Can each developer only have a max number of segments before the app starts to have trouble and, therefore, the developer should delete segments that will not be used soon or make sure they never add more than X segments at a time? I think that would be acceptable, it just needs to be well documented so that developers can adjust their workflows. For example, I might think it's convenient to create all the segments I will need for the next 3 months, but if I know that's too much, I can add only what I need for the week, use it in a batch, and when the annotators are finished I delete everything and start the part.

**Comment by bpadovese** [comment]
Yes, Just to be clear, when I was generating the 50000+ segments, I was really trying to push the limits of the annotation tool. It isnt part of my workflow. However, I agree with Fabio, if there are constraints, I think those need to be made clear both in the instructions/documentation as well as at least warning messages for when the user tries to do something they are not supposed to. For instance, the first model I trained had a couple thousands detentions because it was bad. If I imported those annotations into HALLO it would already create several thousands of segments. If the tool throws an error or warning message, than at least I can upload my annotations in smaller batches.

Just adding to point 7

7 - I have a question here, What if instead of a batch with 4000 segments, I had 40 batches with 100? Would I run into the same problem? Then the same caution from point 6 would apply here.
